README
= Camping, a Microframework

Camping is a web framework which consistently stays at less than 4kB of code.
You can probably view the complete source code on a single page. But, you
know, it's so small that, if you think about it, what can it really do?

The idea here is to store a complete fledgling web application in a single
file like many small CGIs. But to organize it as a Model-View-Controller
application like Rails does. You can then easily move it to Rails once you've
got it going.

== A Camping Skeleton

A skeletal Camping blog could look like this:

  require 'camping'

  Camping.goes :Blog

  module Blog::Models
    class Post < Base; belongs_to :user; end
    class Comment < Base; belongs_to :user; end
    class User < Base; end
  end

  module Blog::Controllers
    class Index
      def get
        @posts = Post.find :all
        render :index
      end
    end
  end

  module Blog::Views
    def layout
      html do
        head { title "My Blog" }
        body do
          h1 "My Blog"
          self << yield
        end
      end
    end

    def index
      @posts.each do |post|
        h1 post.title
      end
    end
  end

== Installation

Interested yet?  Luckily it's quite easy to install Camping.  We'll be using
a tool called RubyGems, so if you don't have that installed yet, go grab it!
Once that's sorted out, open up a Terminal or Command Line and enter:

    gem install camping

Even better, install the Camping Omnibus, a full package of recommended libs:

    gem install camping-omnibus --source http://gems.judofyr.net

If not, you should be aware of that Camping itself only depends on
Rack[http://rack.rubyforge.org], and if you're going to use the views you also
need to install +markaby+, and if you're going to use the database you need
+activerecord+ as well.

    gem install markaby
    gem install activerecord

== Learning

First of all, you should read {the first chapters}[link:book/01_introduction.html]
of The Camping Book. It should hopefully get you started pretty quick. While
you're doing that, you should be aware of the _reference_ which contains
documentation for all the different parts of Camping.

{The wiki}[http://wiki.github.com/camping/camping] is the place for all tiny,
useful tricks that we've collected over the years.  Don't be afraid to share
your own discoveries; the more, the better!

And if there's anything you're wondering about, don't be shy, but rather
subscribe to {the mailing list}[http://rubyforge.org/mailman/listinfo/camping-list]
and ask there.  We also have an IRC channel over at Freenode, so if you feel
like chatting with us, you should join {#camping @ irc.freenode.net}[http://java.freenode.net/?channel=camping].

== Authors

Camping was originally crafted by {why the lucky stiff}[http://en.wikipedia.org/wiki/Why_the_lucky_stiff],
but is now maintained by the _community_.  This simply means that if we like your
patch, it will be applied.  Everything is managed through {the mailing list}[http://rubyforge.org/mailman/listinfo/camping-list],
so just subscribe and you can instantly take a part in shaping Camping.


CHANGELOG
= 2.0
=== 9th Apr, 2010
* Speed-up of Camping::Mab (thanks zimbatm!)
* @state is now an alias of @env['rack.session']
* Camping.use injects a Rack middleware.
* Update Flipbook to RDoc 2.4
* Removed old examples.
* Updated examples/blog.rb
* Camping::Apps returns!
* Session-cookies now timeout naturally (thanks jenna!)
* You can now `throw :halt` to halt the response in a helper.
* Camping::H#u is gone (was an alias to merge!)
* Camping::Session now uses session-cookies. The AR-backend is gone for now.
* camping/db.rb has been renamed to camping/ar.rb.
* Camping now uses Rack internally. Every app responds to #call.

= 1.6
=== Never released

* Camping::Apps removed, it wasn't reliable.
* bin/camping server kinds splitted in various files.
* NotFound and ServerError controllers changed to methods :

  r404 : called when a controller was not found
  r500 : called on uncaught exception
  r501 : called on undefined method

  All of those can be overridden at your taste.

* Markaby no longer required. Like AR, is it autoloaded on (Mab) usage.
* Camping::H is now inheriting from Hash instead of HashWithIndifferentAccess.
* Which made possible to remove the last strict dependency : active_support
* #errors_for removed, it wasn't really used
* Bug fixes !

= 1.5
=== 3rd Oct, 2006

* Camping::Apps stores an array of classes for all loaded apps.
* bin/camping can be given a directory. Like: <tt>camping examples/</tt>
* Console mode -- thank zimbatm. Use: camping -C yourapp.rb
* Call controllers with Camping.method_missing.

    Tepee.get(:Index) #=> (Response)
    Blog.post(:Delete, id) #=> (Response)

    Blog.post(:Login, :input => {'username' => 'admin', 'password' => 'camping'})
      #=> #<Blog::Controllers::Login @user=... >

    Blog.get(:Info, :env => {:HTTP_HOST => 'wagon'})
      #=> #<Blog::Controllers::Info @env={'HTTP_HOST'=>'wagon'} ...>

* Using \r\n instead of \n on output. FastCGI has these needs.
* ActiveRecord no longer required or installed.
* If you refer to Models::Base, however, ActiveRecord will be loaded with autoload. (see lib/camping/db.rb)
* new Camping::FastCGI.serve which will serve a whole directory of apps
  (see http://code.whytheluckystiff.net/camping/wiki/TheCampingServer)
* ~/.campingrc can contain database connection info if you want your default to be something other than SQLite.

    database:
      adapter: mysql
      username: camping
      socket: /tmp/mysql.sock
      password: NOFORESTFIRES
      database: camping

* controllers are now *ordered*. uses the inherited hook to keep track of all
  classes created with R. those classes are scanned, in order, when a request is
  made. any other controllers are handled first. so if you plan on overriding the
  urls method, be sure to subclass from R().
* Console mode will load .irbrc in the working directory, if present.
  (for example, in my ~/git/balloon directory, i have this in the .irbrc:
  include Balloon::Models
  when camping -C balloon.rb gets run, the models all get included in main.)
* And, of course, many other bugfixes from myself and the loyal+kind zimbatm...
* Markaby updated to 0.5. (See its CHANGELOG.)

= 1.4.2
=== 18th May, 2006

* Efficient file uploads for multipart/form-data POSTs.
* Camping tool now uses Mongrel, if available.  If not, sticks with WEBrick.
* Multiple apps can be loaded with the camping tool, each mounted according to their file name.

= 1.4.1
=== 3rd May, 2006

* Streaming HTTP support.  If body is IO, will simply pass to the controller.  Mongrel, in particular, supports this nicely.

= 1.4
=== 11th April, 2006

* Moved Camping::Controllers::Base to Camping::Base.
* Moved Camping::Controllers::R to Camping::R.
* New session library (lib/camping/session.rb).
* WEBrick handler (lib/camping/webrick.rb) and Mongrel handler (lib/camping/mongrel.rb).
* Helpers#URL, builds a complete URL for a route.  Returns a URI object. This way relative links could just return self.URL.path.
* Base#initialize takes over some of Base#service's duties.
* ENV now available as @env in controllers and views.
* Beautiful multi-page docs without frames!

= 1.3
=== 28th January, 2006

* bin/camping: an application launcher.
* <tt>Camping.run(request, response)</tt> now changed to <tt>controller = Camping.run(request, env)</tt>
* This means outputting the response is the wrapper/server's job.  See bin/camping, you can do a controller.to_s at the least.
* <tt>Controllers::Base.env</tt> is the new thread-safe home for <tt>ENV</tt>.
* The input hash now works more like Rails params.  You can call keys
  like methods or with symbols or strings.
* Queries are now parsed more like PHP/Rails, in that you can denote
  structure with brackets: post[user]=_why;post[id]=2
* Auto-prefix table names, to help prevent name clash.
* Helpers.errors_for simple validation.
* Lots of empty :href and :action attributes, a bug.
* New single-page flipbook RDoc template.

= 1.2
=== 23rd January, 2006

* Camping.goes allows fresh modules build from all Camping parts.
* File uploads now supported (multipart/form-data).
* Helpers.R can rebuild routes.
* Helpers./ for tracing paths from the root.

= 1.1
=== 19th January, 2006

* Allowed request and response streams to be passed in, to allow WEBrick and FastCGI support.

= 1.0
=== 17th January, 2006

* Initial checkin, see announcement at http://redhanded.hobix.com/bits/campingAMicroframework.html.


examples/migratoryblog.rb
#!/usr/bin/env ruby

$:.unshift File.dirname(__FILE__) + '/../lib'
require 'camping'
require 'camping/ar'
require 'camping/session'
require 'redcloth'

Camping.goes :MigratoryBlog

module MigratoryBlog
  include Camping::Session

  module Models
    class Post < Base
      belongs_to :user
      
      before_save do |record|
        cloth = RedCloth.new(record.body)
        cloth.hard_breaks = false
        record.html_body = cloth.to_html
      end
    end
    
    class Comment < Base; belongs_to :user; end
    class User < Base; end

    class BasicFields < V 1.1
      def self.up
        create_table :migratoryblog_posts, :force => true do |t|
          t.integer :user_id,          :null => false
          t.string  :title,            :limit => 255
          t.text    :body, :html_body
          t.timestamps 
        end
        create_table :migratoryblog_users, :force => true do |t|
          t.string  :username, :password
        end
        create_table :migratoryblog_comments, :force => true do |t|
          t.integer :post_id,          :null => false
          t.string  :username
          t.text    :body, :html_body
          t.timestamps
        end
        User.create :username => 'admin', :password => 'camping'
      end
      
      def self.down
        drop_table :migratoryblog_posts
        drop_table :migratoryblog_users
        drop_table :migratoryblog_comments
      end
    end
  end

  module Controllers
    class Index
      def get
        @posts = Post.all(:order => 'updated_at DESC')
        render :index
      end
    end

    class PostNew
      def get
        require_login!
        @post = Post.new
        render :add
      end
      
      def post
        require_login!
        post = Post.create(:title => input.post_title, :body => input.post_body,
          :user_id => @state.user_id)
        redirect PostN, post
      end
    end

    class PostN
      def get(post_id) 
        @post = Post.find(post_id)
        render :view
      end
    end

    class Edit < R '/post/(\d+)/edit'
      def get(post_id)
        require_login!
        @post = Post.find(post_id)
        render :edit
      end

      def post(post_id)
        require_login!
        @post = Post.find(post_id)
        @post.update_attributes :title => input.post_title, :body => input.post_body
        redirect PostN, @post  
      end
    end

    class Login
      def get
        render :login
      end
      
      def post
        @user = User.find_by_username_and_password(input.username, input.password)

        if @user
          @state.user_id = @user.id
          redirect R(Index)
        else
          @info = 'Wrong username or password.'
        end
        
        render :login
      end
    end

    class Logout
      def get
        @state.user_id = nil
        redirect Index
      end
    end

    class Style < R '/styles\.css'
      STYLE = File.read(__FILE__).gsub(/.*__END__/m, '')

      def get
        @headers['Content-Type'] = 'text/css; charset=utf-8'
        STYLE
      end
    end
  end
  
  module Helpers
    def logged_in?
      !!@state.user_id
    end
    
    def require_login!
      unless logged_in?
        redirect Controllers::Login
        throw :halt
      end
    end
  end

  module Views
    def layout
      html do
        head do
          title 'My Blog'
          link :rel => 'stylesheet', :type => 'text/css', 
          :href => '/styles.css', :media => 'screen'
        end
        body do
          h1 { a 'My Blog', :href => R(Index) }
          
          div.wrapper! do
            text yield
          end
          
          hr
          
          p.footer! do
            if logged_in?
              _admin_menu
            else
              a 'Login', :href => R(Login)
              text ' to the adminpanel'
            end
            text ' &ndash; Powered by '
            a 'Camping', :href => 'http://camping.rubyforge.org/'
          end
        end
      end
    end

    def index
      if @posts.empty?
        h2 'No posts'
        p do
          text 'Could not find any posts. Feel free to '
          a 'add one', :href => R(PostNew)
          text ' yourself. '
        end
      else
        @posts.each do |post|
          _post(post)
        end
      end
    end

    def login
      h2 'Login'
      p.info @info if @info
      
      form :action => R(Login), :method => 'post' do
        input :name => 'to', :type => 'hidden', :value => @to if @to
        
        label 'Username', :for => 'username'
        input :name => 'username', :id => 'username', :type => 'text'

        label 'Password', :for => 'password'
        input :name => 'password', :id => 'password', :type => 'password'

        input :type => 'submit', :class => 'submit', :value => 'Login'
      end
    end

    def add
      _form(@post, :action => R(PostNew))
    end

    def edit
      _form(@post, :action => R(Edit, @post))
    end

    def view
      _post(@post)
    end

    # partials
    def _admin_menu
      text [['Log out', R(Logout)], ['New', R(PostNew)]].map { |name, to|
        capture { a name, :href => to}
      }.join(' &ndash; ')
    end

    def _post(post)
      h2 { a post.title, :href => R(PostN, post) }
      p.info do
        text "Written by <strong>#{post.user.username}</strong> "
        text post.updated_at.strftime('%B %M, %Y @ %H:%M ')
        _post_menu(post)
      end
      text post.html_body
    end
    
    def _post_menu(post)
      if logged_in?
        a '(edit)', :href => R(Edit, post)
      end
    end

    def _form(post, opts)
      form({:method => 'post'}.merge(opts)) do
        label 'Title', :for => 'post_title'
        input :name => 'post_title', :id => 'post_title', :type => 'text', 
              :value => post.title

        label 'Body', :for => 'post_body'
        textarea post.body, :name => 'post_body', :id => 'post_body'

        input :type => 'hidden', :name => 'post_id', :value => post.id
        input :type => 'submit', :class => 'submit', :value => 'Submit'
      end
    end
  end
  def self.create
    MigratoryBlog::Models.create_schema :assume => (MigratoryBlog::Models::Post.table_exists? ? 1.0 : 0.0)
  end
end

=begin
def MigratoryBlog.create
  MigratoryBlog::Models.create_schema :assume => (MigratoryBlog::Models::Post.table_exists? ? 1.0 : 0.0)
end
=end

__END__
* {
  margin: 0;
  padding: 0;
}

body {
  font: normal 14px Arial, 'Bitstream Vera Sans', Helvetica, sans-serif;
  line-height: 1.5;
}

h1, h2, h3, h4 {
  font-family: Georgia, serif;
  font-weight: normal;
}

h1 {  
  background-color: #EEE;
  border-bottom: 5px solid #6F812D;
  outline: 5px solid #9CB441;       
  font-weight: normal;
  font-size: 3em;  
  padding: 0.5em 0;
  text-align: center;
}

h2 {
  margin-top: 1em;
  font-size: 2em;
}

h1 a { color: #143D55; text-decoration: none }
h1 a:hover { color: #143D55; text-decoration: underline }
h2 a { color: #287AA9; text-decoration: none }
h2 a:hover { color: #287AA9; text-decoration: underline }

#wrapper { 
  margin: 3em auto;
  width: 700px;
}

p {
  margin-bottom: 1em;
}

p.info, p#footer {
  color: #999;
  margin-left: 1em;
}

p.info a, p#footer a {
  color: #999;
}

p.info a:hover, p#footer a:hover {
  text-decoration: none;
}

a {
  color: #6F812D;
}

a:hover {
  color: #9CB441;
}

hr {
  border-width: 5px 0;
  border-style: solid;     
  border-color: #9CB441;
  border-bottom-color: #6F812D;
  height: 0;   
}

p#footer {    
  font-size: 0.9em;
  margin: 0;      
  padding: 1em;
  text-align: center;
}

label {  
  display: block;
  width: 100%;
}

input, textarea {
  padding: 5px;     
  margin-bottom: 1em;
  margin-right: 490px;
  width: 200px;  
}

input.submit {
  width: auto;
}

textarea {
  font: normal 14px Arial, 'Bitstream Vera Sans', Helvetica, sans-serif;
  height: 300px;
  width: 400px;
}

